docs/specs/buffer.md

```diff
@@ -0,0 +1,54 @@
+# Buffer Specification
+
+This document specifies Vector's buffer behavior for the development of Vector.
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”,
+“SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be
+interpreted as described in [RFC 2119].
+
+<!-- MarkdownTOC autolink="true" style="ordered" indent="   " -->
+
+1. [Instrumentation](#instrumentation)
+
+<!-- /MarkdownTOC -->
+
+## Instrumentation
+
+Vector buffers MUST be instrumented for optimal observability and monitoring. This is required to drive various interfaces that Vector users depend on to manage Vector installations in mission critical production environments. This section extends the [Instrumentation Specification].
+
+### Events
+
+#### `EventsReceived`
+
+*All buffers* MUST emit an `EventsReceived` event immediately after receiving one or more Vector events.
+
+* Properties
+  * `count` - the number of received events
+  * `byte_size` - the byte size of received events
+* Metric
+  * MUST increment the `buffer_event_count` gauge by the defined `count`
```

Comment:
Consistent with my comment [here](https://github.com/vectordotdev/vector/pull/9165#issuecomment-921259848) by adjusting this type to gauge we now no longer have the ability to calculate the rate at which items enter the buffer: we're still missing two legs of Little's law. Moving these metrics to gauges mean we can't calculate per second bytes or item rates, even ignoring the Little's law ambition. If we have to have just one metric type a count would be more useful generally. 

Can we add a count and a gauge here? The count would maintain items/bytes seen since startup and the gauge the instantaneous value. We get latency information for free with both of these, plus the ability to calculate rates.

Reply:
A 'gauge' type called `_count` is also sort of weird for contexts where `count` is meant to be a monotonically increasing value and gauges are not.

Reply:
I think we need both a gauge of the current sizes, and a counter on the number and size of events in and out of the buffer.

Reply:
Yep. I'm proposing that for each unit we want a gauge of the instantaneous sum of the unit in the buffer -- so the instantaneous number of items in the buffer, the instantaneous number of bytes -- and a counter of the unit as they enter the buffer. With those in hand we can compute the average latency of the unit through the buffer.

Reply:
👍 yeah these are good points. I can see the value of recording the counts for calculating rates, which users are likely to want to graph. And that average latency would be calculable given those other metrics.

Reply:
Apologies, I was meaning to add a note about the omission of `received_events_total`. I'd removed it after discussion elsewhere that we may prefer to track it internally and emit a latency metric directly in the future. Will add the counters back in.

Good point on the naming; perhaps `buffer_events` would be better?

Reply:
> +1 yeah these are good points. I can see the value of recording the counts for calculating rates, which users are likely to want to graph. And that average latency would be calculable given those other metrics.

Yep, we have use in our performance work for both having the rate and knowing what the latency through the buffer is. 

> Apologies, I was meaning to add a note about the omission of received_events_total. I'd removed it after discussion elsewhere that we may prefer to track it internally and emit a latency metric directly in the future. Will add the counters back in.

Tracking latency directly isn't necessary if we have the two other legs of Little's law, the benefit to tracking those two other legs being counters are cheaper than clocks. We don't get distribution, of course.